### PR TITLE
Fixes #31425 - Render statuses

### PR DESCRIPTION
**Pull request by kamils-iRonin**
Reopened https://github.com/theforeman/foreman/pull/8377

I have a feeling the previous PR was overcomplicated and I wanted to keep this one as simple as possible. It includes saving the status after each render attempt and displaying that on the template and host pages. I think this is more about the template than the host (we want to make sure that the templates are not broken), so I'm not really sure if host status is suitable here. Once this is merged, we can think about some kind of monitoring/notifications.

Here are some screenshots:

image
image
image

### CODE REVIEW

**Comment by amirfefer** [comment]
Thanks @kamils-iRonin!

I'm not sure if the card "Render Statuses" fits the overview tab, it has too much data IMHO which takes much of the screen's real estate, the overview tab mostly shows recent / aggregated information regarding the host.

**Comment by MariSvirik** [comment]
Hey, @kamils-iRonin it's a great effort. 
I agree with what @amirfefer said.
Still, I have a couple of questions about render statuses as I'm hearing about it for the first time. (Apologies for a lack of my technical knowledge)
What are render statuses? Do they tell a user if a template rendered correctly?
How is it important for a user? And is it important in the context of a specific host?

**Comment by kamils-iRonin** [comment]
@amirfefer @MariSvirik thanks for the feedback! I'll try to address the questions below, do not hesitate to ask if something is still unclear 😄 

> What are render statuses?

Basically, the render statuses are designed to provide information about possible rendering problems. We would like to know which templates are used with which hosts and if any errors occurred during the rendering process.

> Do they tell a user if a template rendered correctly? How is it important for a user?

Exactly, with this we can prevent further problems, if we detect that there is a problem with rendering some template, we can check which hosts may be affected. Also, if you use the renderer in unsafe mode, it can be helpful to migrate to safe mode.

> And is it important in the context of a specific host?

The first idea was to introduce a new host status, but now I think it's more about the template. I don't have a strong opinion on whether the render statuses should be displayed on the host page or not.

> I'm not sure if the card "Render Statuses" fits the overview tab, it has too much data IMHO which takes much of the screen's real estate.

Maybe it would be enough to show how many templates were rendered for this particular host and if all renderings were successful? (we can also add a "Details" button)

**Comment by ares** [comment]
Great work so far! I'd say the templates would deserve it's own tab. Either for this table with perhaps more info (link to audits of this template, info about it's type etc) or it should be a part of a new Provisioning tab. We can start with Templates tab and then perhaps move it if it would be better. Just my 2 cents

**Comment by Ron-Lavi** [comment]
Hey @kamils-iRonin what is the status here?
is it just waiting for another review?

**Comment by kamils-iRonin** [comment]
@Ron-Lavi If I remember correctly, there are minor things to improve and tests to fix. I'm quite busy with another project at the moment, but I'll try to finish it in the next few days.

**Comment by kamils-iRonin** [comment]
I believe the failed JavaScript tests are not related, I just ran the tests on `develop` and I see the same errors

**Comment by kamils-iRonin** [comment]
[test foreman]

**Comment by kamils-iRonin** [comment]
@Ron-Lavi I think it's ready for another review

**Comment by Ron-Lavi** [comment]
Hey @kamils-iRonin, I guess this PR fell through the cracks,
I rebased it for you and fixed the conflict with the [safemode UI setting change](https://github.com/theforeman/foreman/blame/627118e65fa0cd67628696a748fbb72498afff74/app/helpers/application_helper.rb#L399)

**Comment by kamils-iRonin** [comment]
Hi @Ron-Lavi, thanks and I think it is ready now

**Comment by sbernhard** [comment]
What shall we do with this PR? Rebase and finish or close it? @MariaAga ?

**Comment by kamils-iRonin** [comment]
It is quite old PR and the details are not entirely fresh in my memory. Just in case I have rebased and tested. It seems to work fine.

**Comment by kmalyjur** [comment]
Since this PR is quite old, I just want to let you know that I'm going to take a look at it within the next few days.

File: app/views/api/v2/render_statuses/base.json.rabl
Comment (kmalyjur) [inline app/views/api/v2/render_statuses/base.json.rabl:3]: This only seems to be used to extend the main.json.rabl. Does it need to be a separate file?

File: webpack/assets/javascripts/react_app/components/RenderStatuses/Empty.js
Comment (kmalyjur) [inline webpack/assets/javascripts/react_app/components/RenderStatuses/Empty.js:13]: We already have `/common/EmptyState` in Foreman, couldn't you please use that instead of creating a new component?

File: webpack/assets/javascripts/react_app/components/RenderStatuses/index.js
Comment (kmalyjur) [inline webpack/assets/javascripts/react_app/components/RenderStatuses/index.js:141]: I don't think we use `widgetId`

File: webpack/assets/javascripts/react_app/components/RenderStatuses/index.js
Comment (kmalyjur) [inline webpack/assets/javascripts/react_app/components/RenderStatuses/index.js:94]: Please import colors like this:
`import { global_success_color_100 as successedColor } from '@patternfly/react-tokens';`
And then you can use them this way:
`<CheckCircleIcon color={successedColor.value} />`

File: webpack/assets/javascripts/react_app/components/RenderStatuses/index.js
Comment (kmalyjur) [inline webpack/assets/javascripts/react_app/components/RenderStatuses/index.js:47]: There is some issue with the labels. It looks like this when the screen is too small:
Screenshot from 2024-03-14 14-31-38

**Review by kmalyjur — CHANGES_REQUESTED** [review]
@adamruzicka helped me with the review (mainly with the backend) and we came to the conclusion that it might be worth adding the same "Render Statuses" tab to the Host Groups and also to the Provisioning templates card in the new host detail page -> Details (sort of like the Overview -> Recent jobs card), however, I know that this has already been discussed and it's possible it was decided not to include it there.
Idk about the tests, hopefully rebasing would help.

**Pull request closed without merging**
